[WFCORE-4750] Introduce regex mapper for security roles in Elytron

Skyllarr opened the pull request:
Proposal for https://issues.jboss.org/browse/WFCORE-4750

darranl commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,70 @@
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1344[EAP7-1344]
```
Maybe not ready yet but I suspect this will also need an ELY issue for tracking.

Skyllarr replied:
Fixed

darranl commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,70 @@
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+[x] Engineering
```
Plan to add tests with everything you do but we may still want to end up with QE testing before merge as this has a reasonable priority.

Skyllarr replied:
Fixed

fabiobrz commented:
Hi @Skyllarr, I am going to start review process on your proposal. 
Can you please add me as `QE contact` and resolve all previously addressed comments so that we can start a new round?

Skyllarr commented:
Hi @fabiobrz . I added you as QE contact and also changed Testing By from Engineering to QE. This is to reflect the fact that EAP7 issue has Testing By also set to QE, since this issue has reasonable priority. However I implemented tests acording to test plan and these are included in the PRs.

fabiobrz commented:
Hi @Skyllarr, thanks, I'll start the review process as soon as possible, although probably tomorrow.

Meantime, regarding what you stated about the `Testing By` section, I am not actually sure...
I mean that - although of course I am aware of the fact that QE is providing review and supervision for test related tasks and double check for pre-checked - IIUC `Testing By` in the proposal document and in the related JIRA issue should be set to `Engineering` when test plan and test development is provided by Engineering.
 
What you say about the RFE ranking is right too, but I see that this is "not so high priority" as it is ranked 31, if I don't go wrong. Am I missing something?

Skyllarr commented:
@fabiobrz Thank you. You are right about Testing by.  I switched both jira and proposal to Engineering. Thanks

fabiobrz commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,109 @@
+
+== Test Plan
+
+* Functionality tests will be added to Elytron and wildfly-core.
```
Please improve this list with a minimal description or list that would let the reader easily understand which projects' test suites will be affected, maybe referring to the `Affected Projects or Components` section - e.g.:

* Elytron test suite: functional tests
* Wildfly Core test suite: smoke tests  
* Wildfly test suite: ...

fabiobrz replied:
I see that you didn't include both [1] and [2] sections from the design template. Is there any reason for this? 
I think that - as you provided a tracker for Wildfly documentation, then this should be mentioned in [1] and also that this nice new feature could be worth at least a sentence in the `Release Notes`. At least I - as a developer  - would find it really helpful if it was there :) 
WDYT?

[1]
https://github.com/wildfly/wildfly-proposals/blob/master/design-doc-template.adoc#community-documentation

[2]
https://github.com/wildfly/wildfly-proposals/blob/master/design-doc-template.adoc#release-note-content

Skyllarr commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,109 @@
+
+== Overview
+
+This task is is about introducing the possibility of role mapping based on regex in Elytron. With this functionality it will be possible for customers to easily translate list of roles (eg. *-admin, *-user) to simpler roles (eg. admin, user) without having to implement their own custom components to achieve this.
```
Thanks for review. I updated the Overview

Skyllarr commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,109 @@
+
+* https://issues.jboss.org/browse/WFCORE-4750[WFCORE-4750]
+
+=== Related Issues
```
Updated, thanks.

Skyllarr commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,109 @@
+
+=== Hard Requirements
+
+It must be possible to use regex for mapping of roles in Elytron. This means user must be able to define pattern in the subsystem that will be used to match roles, and also define a string that these roles should be mapped to. Role matches the pattern iff the pattern can be found in the role.
```
> Role matches the pattern iff the pattern can be found in the role.
> Is this sentence really necessary? Aren't the concept of regular expression and the description you're providing enough?

@fabiobrz This sentence should have been better explained by me. What I meant is that not whole `role` string has to match in order for the role to be matched (difference between methods [find()](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Matcher.html#find()) and [matches()](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Matcher.html#matches()) on Java class [Matcher](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Matcher.html) ). This design also allows user to specify whether to replace all instances or only the first one ( see `replace-all` option discussed in this proposal). This way we are more consistent with `RegexNameRewriter` which is another Elytron mapper that has option `replace-all`.  It is still possible to require whole text match with the use of special regex symbols `^` at the beginning and `$` at the end. But this decision can be discussed and changed if some other option is better.

Skyllarr commented:
@fabiobrz Thank you for review. I updated the analysis document and took into consideration all your comments. Let me know if something is still unclear.

fabiobrz commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,159 @@
+
+
+ * In the example below the domain part is extracted from input roles that are in form of an email address in order to convert them to the `{extracted_domain}-role` format.
+E.g.: `user@gmail.com` and `user@customApp.com` input roles would result in `gmail-role` and `customerApp-role` security roles).
```
Typo: remove the dangling `)` here - just after the word `roles`.

fabiobrz commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,159 @@
+== Test Plan
+
+* WildFly Elytron test suite: Functional tests  - tests for regex role mapper functionality
+* WildFly Core tess suite: Functional tests - testing of regex role mapper functionality when it is defined in the subsystem
```
Typo: `WildFly Core tess...` -> `WildFly Core test...`

fabiobrz commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,159 @@
+== Test Plan
+
+* WildFly Elytron test suite: Functional tests  - tests for regex role mapper functionality
+* WildFly Core tess suite: Functional tests - testing of regex role mapper functionality when it is defined in the subsystem
```
So the Wildfly Test Suite will not be involved at all?

Skyllarr replied:
@fabiobrz No the WildFly test suite will not be involved. Should I remove it from affected components to make it clear? Purpose of adding WildFly Security there was because it introduces changes to the Elytron subsystem in the WildFly server, but maybe it is confusing.

fabiobrz replied:
No, it makes definitely sense to have it there. This was just a question that I asked for clarification. Thanks.

fabiobrz commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,159 @@
+
+== Community Documentation
+
+WildFly - community documentation will be delivered in form of a PR to master branch.
```
Could you add a pointer like `as detailed in the "Related Issues" section?

Skyllarr replied:
Fixed

fabiobrz commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,159 @@
+<regex-role-mapper name="my-rrm" pattern=".*@([a-z]*)\..*" replacement="$1-role" keep-non-mapped="true"/>
+----
+
+ * The following role mapper can be used to replace all instances of the `guest` substring with the` user` substring.
```
Typo: 
```
substring with the` user`
```
->
```
substring with the `user`
```

Skyllarr replied:
Fixed

fabiobrz commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,159 @@
+----
+
+ * The following role mapper can be used to replace all instances of the `guest` substring with the` user` substring.
+E.g. the `abc-guest-abc-guest` input role would be mapped to the` abc-user-abc-user` security role.
```
Typo: 
```
the` abc-user-abc-user`
```
->
```
the `abc-user-abc-user`
```

Skyllarr replied:
Fixed

fabiobrz commented on elytron/WFCORE-4750-regex-role-mapper.adoc:
```diff
@@ -0,0 +1,159 @@
+<regex-role-mapper name="my-rrm" pattern=".*@([a-z]*)\..*" replacement="$1-role" keep-non-mapped="true"/>
+----
+
+ * The following role mapper can be used to replace all instances of the `guest` substring with the`user` substring.
```
“the‘user’” -> “the ‘user’” :-)

fabiobrz commented:
Hi @Skillar, there’s a small typo that I commented on above but I approved the PR.
If you can please fix it, thanks for your grate work on this!

Skyllarr commented:
> Hi @skyllar, there’s a small typo that I commented on above but I approved the PR.
> If you can please fix it, thanks for your grate work on this!

Fixed. Thanks as well
